tuist/tuist
# 🚧 Add Multi Platform support for Targets

Resolves https://github.com/tuist/tuist/issues/397

### Short description 📝

Following the blog post from Dave De Long (https://davedelong.com/blog/2018/11/15/building-a-crossplatform-framework/) we are adding support into the manifest for defining a target which supports multiple platforms.

```swift
import ProjectDescription

let project = Project(name: "Framework2",
                      targets: [
                          Target(name: "Framework2",
                                 platform: [ .iOS, .macOS ],
                                 product: .framework,
                                 bundleId: "io.tuist.Framework2",
                                 infoPlist: "Config/Framework2-Info.plist",
                                 sources: "Sources/**",
                                 headers: Headers(public: "Sources/Public/**", 
                                                  private: "Sources/Private/**", 
                                                  project: "Sources/Project/**"),
                                 dependencies: []),

                          Target(name: "Framework2Tests",
                                 platform: [ .iOS, .macOS ],
                                 product: .unitTests,
                                 bundleId: "io.tuist.Framework2Tests",
                                 infoPlist: "Config/Framework2Tests-Info.plist",
                                 sources: "Tests/**",
                                 dependencies: [
                                     .target(name: "Framework2"),
                          ]),
])
```

### Solution 📦

- Use an OptionSet in the manifest to allow _one_ or _many_ platforms to be specified, this also means it's impossible to define duplicates.
- When generating the project settings if multiple platforms are specified ensure build settings include the `[sdk=_*]` syntax. e.g.

```
        "SWIFT_ACTIVE_COMPILATION_CONDITIONS[sdk=macosx*]": "DEBUG",
        "CODE_SIGN_IDENTITY[sdk=iphoneos*]": "iPhone Developer",
        "SDKROOT[sdk=macosx*]": "macosx",
        "COMBINE_HIDPI_IMAGES[sdk=macosx*]": "YES",
```

- Ensure the linter now checks that _any_ of the supported platforms can resolve the dependency graph.

### Implementation 👩‍💻👨‍💻

- [x] Add manifest API for platform
- [x] Change build settings per-platform
- [x] Linter for multi platform targets 

### Test Plan 🛠:

- Generate `fixtures/ios_app_with_frameworks` via `tuist generate`
- Run the following:
```
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk macosx
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk macosx
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk iphoneos
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk iphoneos
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk iphonesimulator
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk iphonesimulator
```
- Ensure _all_ builds succeed
_optional_:
- Open the project in Xcode 
- Play with it
